On theme switch, delete the previous if it was a wpcom theme

eliorivero opened the pull request:
#### Changes proposed in this Pull Request:

* during a theme switch, check if the theme previously active was a wpcom premium theme and delete it if that was the case
* new method `Jetpack::is_premium_theme` that checks if a theme is a wpcom premium theme

#### Testing instructions:

0. mock a wpcom premium theme like Mood and name the folder `mood-wpcom` because that's the name that they get when they're downloaded to a Jetpack site.
1. in WP Admin > Appearance, activate the Mood theme
2. now activate a free wpcom theme like Dara. You can either mock it or do this from Calypso
3. verify that Mood theme was deleted
4. in WP Admin > Appearance, activate any other theme like TwentySeventeen
5. verify that Dara theme was not deleted

dereksmart commented on json-endpoints/jetpack/class.jetpack-json-api-themes-active-endpoint.php:
```diff
@@ -38,8 +38,14 @@ protected function switch_theme() {
+
 		switch_theme( $theme_slug );
 
+		if ( isset( $current_theme['id'] ) && $this->is_wpcom_theme( $current_theme['id'] ) ) {
```
Will this also delete free wpcom themes?

roccotripaldi replied:
yes. is that okay?

dereksmart replied:
Yeah, just feels a little weird so want to make sure it's a conscious decision.

dereksmart commented:
Works well from Calypso.  Do you think we should be hooking into this if the theme was switched from within wp-admin as well?

dereksmart commented:
Whoops, didn't mean to close

dereksmart commented:
> What if after switching, they decide to switch back?

@roccotripaldi I figured that since there is a reason we're deleting these when switching from Calypso, that the same reason would be relevant when switching from anywhere else.  

@eliorivero, thoughts?

eliorivero commented:
@dereksmart @roccotripaldi
Yes, I need update this so the theme is deleted when it's switched in WP Admin as well.

eliorivero commented:
Updated to hook into `switch_theme`. With this the code in the json endpoint is no longer necessary. Covers cases when theme is switched from Calypso, WP Admin or wp-cli.

roccotripaldi commented:
> I just thought of something: If someone created a child theme off of a wpcom theme, and then switches from the wpcom theme to the child theme, the parent would be deleted.

Oof. Good catch.

eliorivero commented:
Updated to add method that checks if a theme is a premium wpcom theme. This method will be used to check if the theme was a wpcom premium theme and only delete it if it is.

eliorivero commented:
Updated to add method that checks if a theme is a premium wpcom theme. This method will be used to check if the theme was a wpcom premium theme and only delete it if it is.

I've also updated the testing instructions.

roccotripaldi commented:
Looking good, and tests well. Should we verify the old theme is not the parent of the current theme before deleting?

eliorivero commented:
Nope, we won't verify that. We definitely need to delete it.

roccotripaldi commented:
This has my thumbs up! 👍

roccotripaldi commented:
I guess an easy way to handle theme deletion on multi-site would be to not delete the theme if we are on multi-site!

Otherwise, we'll need to loop through each site on the network and check if it is using the theme before we delete it. 

I do not think we support premium theme installation on multi-site networks yet. So maybe this is a future consideration.
